pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,505 @@
+	cfgBaseApp.P2P.MaxPeers = int(cfg.P2P.MaxPeers)
+
+	err = cfgBase.ProtocolConfiguration.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid composed protocol configuration: %w", err)
+	}
+
+	cfgServer, err := network.NewServerConfig(cfgBase)
+	if err != nil {
+		return nil, fmt.Errorf("compose NeoGo server config from the base one: %w", err)
+	}
+
+	var cfgDB dbconfig.DBConfiguration
+	cfgDB.Type = cfg.Storage.typ
+	if cfgDB.Type == dbconfig.BoltDB {
+		cfgDB.BoltDBOptions.FilePath = cfg.Storage.path
+	} else if cfgDB.Type == dbconfig.LevelDB {
+		cfgDB.LevelDBOptions.DataDirectoryPath = cfg.Storage.path
+	}
+
+	bcStorage, err := storage.NewStore(cfgDB)
+	if err != nil {
+		return nil, fmt.Errorf("init storage for blockchain: %w", err)
+	}
+
+	defer func() {
+		if err != nil {
+			closeErr := bcStorage.Close()
+			if closeErr != nil {
+				err = fmt.Errorf("%w; also failed to close blockchain storage: %v", err, closeErr)
+			}
+		}
+	}()
+
+	bc, err := core.NewBlockchain(bcStorage, cfgBase.Blockchain(), cfg.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("init core blockchain component: %w", err)
+	}
+
+	netServer, err := network.NewServer(cfgServer, bc, bc.GetStateSyncModule(), cfg.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("init NeoGo network server: %w", err)
+	}
+
+	if !cfg.NotaryDisabled {
+		var cfgNotary notary.Config
+		cfgNotary.MainCfg.Enabled = true
+		cfgNotary.MainCfg.UnlockWallet = cfg.Wallet
+		cfgNotary.Chain = bc
+		cfgNotary.Log = cfg.Logger
+
+		notaryService, err := notary.NewNotary(cfgNotary, netServer.Net, netServer.GetNotaryPool(), func(tx *transaction.Transaction) error {
+			err := netServer.RelayTxn(tx)
+			if err != nil && !errors.Is(err, core.ErrAlreadyExists) {
+				return fmt.Errorf("relay completed notary transaction %s: %w", tx.Hash().StringLE(), err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init Notary service: %w", err)
+		}
+
+		netServer.AddService(notaryService)
+		bc.SetNotary(notaryService)
+	}
+
+	var cfgConsensus consensus.Config
+	cfgConsensus.Logger = cfg.Logger
+	cfgConsensus.Broadcast = netServer.BroadcastExtensible
+	cfgConsensus.Chain = bc
+	cfgConsensus.BlockQueue = netServer.GetBlockQueue()
+	cfgConsensus.ProtocolConfiguration = bc.GetConfig().ProtocolConfiguration
+	cfgConsensus.RequestTx = netServer.RequestTx
+	cfgConsensus.StopTxFlow = netServer.StopTxFlow
+	cfgConsensus.TimePerBlock = cfg.BlockInterval
+	cfgConsensus.Wallet = cfg.Wallet
+
+	consensusService, err := consensus.NewService(cfgConsensus)
+	if err != nil {
+		return nil, fmt.Errorf("init Consensus service: %w", err)
+	}
+
+	netServer.AddConsensusService(consensusService, consensusService.OnPayload, consensusService.OnTransaction)
+
+	// "make" channel rw to satisfy Start method
+	chErrRw := make(chan error)
+
+	go func() {
+		for {
+			err, ok := <-chErrRw
+			if !ok {
+				return
+			}
+			cfg.ErrorListener <- err
+		}
```

Comment:
How about exiting this loop on IR stop and draining/closing the channel afterwards? It's not so vital, but still.

Reply:
I thought about it. We provide writable channel to `rpcsrv.Server`, so technically we shouldn't close it to prevent potential panic (which is not critical problem, but not a good practice). Relying on the current design of the components, it is expected that when `Stop` is called, the channel will be closed inside the components based on this channel, and when it is closed, we exit the routine.

What do you think?

Reply:
> technically we shouldn't close it to prevent potential panic

It's a subject for discussion, but technically, we should firstly wait for the service to finish its operation (we have `rpcServer.Shutdown()` for this, it returns after service is completely closed) and then close the channel.

> the channel will be closed inside the components based on this channel

How about several services based on a single error channel? One service may be shutted down, but the others should still be able to use the provided channel. We have it in neo-go, BTW. 

In general, I think that the caller is resposible for resources release, but in some urgent and special cases the callee may be allowed to close the channel (like RPC client does with the subscriber's channel when it receives MissedEvent). I don't think it's a spacial case.

@roman-khimov, who is responsible for channel's closing, the caller or the callee?

Reply:
It's all about `chErrRw`.

Reply:
> How about several services based on a single error channel?

IMO external channel should never be closed, otherwise `panic: send on closed channel` occurs. We should close only channels which are completely controlled by the single layer of code.

Reply:
@roman-khimov thoughts?